File: tests/data-conversion/class-generic/genericla.cpp
#include "runtime/array.h"
#include "runtime/union.h"

class Item;
template<typename T, typename U>
class Wrapper;
void TestGenericClass();

class Item : public compilets::Object {
};

template<typename T, typename U>
class Wrapper : public compilets::Object {
 public:
  compilets::CppgcMemberType<T> member;
  compilets::OptionalCppgcMemberType<T> optionalMember;
  compilets::Union<compilets::CppgcMemberType<T>, compilets::CppgcMemberType<U>> unionMember;
  compilets::Union<compilets::CppgcMemberType<T>, compilets::CppgcMemberType<U>, std::monostate> optionalUnionMember;
  cppgc::Member<compilets::Array<compilets::CppgcMemberType<T>>> arrayMember = compilets::MakeArray<compilets::CppgcMemberType<T>>({});

  virtual void method() {
    compilets::ValueType<T> m = this->member;
    m = compilets::GetOptionalValue(this->optionalMember);
    m = std::get<compilets::CppgcMemberType<T>>(this->unionMember);
    m = std::get<compilets::CppgcMemberType<T>>(this->optionalUnionMember);
    m = this->arrayMember->value()[0];
  }

  virtual void take(compilets::ValueType<T> value) {
    this->member = value;
    this->optionalMember = value;
    this->unionMember = value;
    this->optionalUnionMember = value;
    this->arrayMember = compilets::MakeArray<compilets::CppgcMemberType<T>>({value});
  }

  void Trace(cppgc::Visitor* visitor) const override {
    compilets::TracePossibleMember(visitor, member);
    compilets::TracePossibleMember(visitor, optionalMember);
    compilets::TracePossibleMember(visitor, unionMember);
    compilets::TracePossibleMember(visitor, optionalUnionMember);
    compilets::TraceMember(visitor, arrayMember);
  }

  virtual ~Wrapper() = default;
};

void TestGenericClass() {
  Wrapper<double, bool>* primitive = compilets::MakeObject<Wrapper<double, bool>>();
  primitive->take(123);
  primitive->method();
  double n = primitive->member;
  n = primitive->optionalMember.value();
  n = std::get<double>(primitive->unionMember);
  n = std::get<double>(primitive->optionalUnionMember);
  n = primitive->arrayMember->value()[0];
  primitive->take(n);
  std::optional<double> optionalNumber = primitive->optionalMember;
  compilets::Union<double, bool> numberOrBool = primitive->unionMember;
  compilets::Union<double, bool, std::monostate> numberOrBoolOrNull = primitive->optionalUnionMember;
  compilets::Array<double>* numberArray = primitive->arrayMember;
  Wrapper<Item, bool>* nested = compilets::MakeObject<Wrapper<Item, bool>>();
  nested->take(compilets::MakeObject<Item>());
  nested->method();
  Item* item = nested->member;
  item = nested->optionalMember;
  item = std::get<cppgc::Member<Item>>(nested->unionMember);
  item = std::get<cppgc::Member<Item>>(nested->optionalUnionMember);
  item = nested->arrayMember->value()[0];
  nested->take(item);
  Item* optionalItem = nested->optionalMember;
  compilets::Union<bool, Item*> itemOrBool = nested->unionMember;
  compilets::Union<bool, Item*, std::monostate> itemOrBoolOrNull = nested->optionalUnionMember;
  compilets::Array<cppgc::Member<Item>>* itemArray = nested->arrayMember;
}

File: tests/data-conversion/class-generic/genericla.ts
class Item {}

class Wrapper<T extends {}, U = boolean> {
  member: T;
  optionalMember?: T;
  unionMember: T | U;
  optionalUnionMember?: T | U;
  arrayMember: T[] = [];

  method() {
    let m = this.member;
    m = this.optionalMember!;
    m = this.unionMember as T;
    m = this.optionalUnionMember as T;
    m = this.arrayMember[0];
  }

  take(value: T) {
    this.member = value;
    this.optionalMember = value;
    this.unionMember = value;
    this.optionalUnionMember = value;
    this.arrayMember = [value];
  }
}

function TestGenericClass() {
  const primitive = new Wrapper<number>;
  primitive.take(123);
  primitive.method();

  let n = primitive.member;
  n = primitive.optionalMember!;
  n = primitive.unionMember as number;
  n = primitive.optionalUnionMember as number;
  n = primitive.arrayMember[0];
  primitive.take(n);

  const optionalNumber = primitive.optionalMember;
  const numberOrBool = primitive.unionMember;
  const numberOrBoolOrNull = primitive.optionalUnionMember;
  const numberArray = primitive.arrayMember;

  const nested = new Wrapper<Item>;
  nested.take(new Item);
  nested.method();

  let item = nested.member;
  item = nested.optionalMember!;
  item = nested.unionMember as Item;
  item = nested.optionalUnionMember as Item;
  item = nested.arrayMember[0];
  nested.take(item);

  const optionalItem = nested.optionalMember;
  const itemOrBool = nested.unionMember;
  const itemOrBoolOrNull = nested.optionalUnionMember;
  const itemArray = nested.arrayMember;
}
